### PR TITLE
summarizeAsync

### LINKED ISSUE #7934: Async Summarization (November)
## Work Item
Completion of #7615 <- see here for full details

**Approach**
The required work for this item is to break the graph visitation into two passes. The first pass visits the IChannels, instructing the channel to capture whatever state it will need in order to summarize itself. Because attachment occurs on the user's client (vs. the designated summarization client), this pass must remain synchronous to avoid local mutations while the snapshot is being captured.

The second pass revisits the same graph of IChannels, requesting that they produce an ISummaryTree from their previously captured state. This second pass is asynchronous, which allows allows the DDS implementer to upload or download blobs, and/or periodically yield.

What the DDS captures during the first pass is up to the DDS. DDSes using functional/persistent/immutable data structures may perform a cheap clone during this phase, while other DDSes may continue to synchronously produce an ITree in the first pass and simply return it in the second pass.

### LINKED ISSUE #8595: Remove usage of ITree in SharedObject.snapshotCore
It would be great to change SharedObject.snapshotCore() API not to deal with ITree, but rather return back ISummaryTreeWithStats and remove this conversion in SharedObject:
```typescript
            const snapshot: ITree = this.snapshotCore(serializer);
            summaryTree = convertToSummaryTreeWithStats(snapshot, fullTree);
```

Basically, this coded was put here to reduce churn when we moved to summaries. It would be great if all layers were talking same language, i.e. all DDSs (based on SharedObject) were actually natively building summaries and not ITree and converting them to summary. Where it's complicated, we can push this conversion one layer deeper (i.e. remove it from SharedObject, but keep some DDS internally doing ITree), but ideally all our "good" DDSs would demonstrate right behavior and use fewer concepts.

As an example RE how to work with summaries (and built ISummaryTreeWithStats), please look at usage of SummaryTreeBuilder() at other layers.

### PR DESCRIPTION
**Pull request by scarlettjlee**
This PR is for parts 1 and 2 of Vlad's comment [here](https://github.com/microsoft/FluidFramework/pull/8422#issuecomment-996919067) on prior abandoned PR.

Add a new method on IChannel (initially optional for compat) to do summarization asynchronously and only call this during regular summarization, not attach.

SharedObject is split into SharedObjectCore and SharedObject, where Core does not have implementations for summarize or GC. DDS implementers are free to implement as they like. For existing DDSs in this repo, summarizeAsync will call into summarizeCore (previously snapshotCore) as synchronous summarize does.

### CODE REVIEW

File: packages/dds/shared-object-base/src/sharedObject.ts [hunk 1]
```diff
@@ -519,3 +447,118 @@ export abstract class SharedObject<TEvent extends ISharedObjectEvents = ISharedO
...
+
+    /**
+     * {@inheritDoc (ISharedObject:interface).summarize}
+     */
+    public summarize(fullTree: boolean = false, trackState: boolean = false): ISummaryTreeWithStats {
+        // Set _isSummarizing to true. This flag is used to ensure that we only use SummarySerializer (created below)
+        // to serialize handles in this object's data. The routes of these serialized handles are outbound routes
+        // to other Fluid objects.
+        assert(!this._isSummarizing, 0x076 /* "Possible re-entrancy! Summary should not already be in progress." */);
+        this._isSummarizing = true;
```
Comment (scarlettjlee) [inline packages/dds/shared-object-base/src/sharedObject.ts:495]: I'm still not convinced that _isSummarizing (and certainly SummarySerializer) is useful during summarize rather than GC-only. Re-entrancy does seem like it shouldn't be happening, but I'm not sure that anything bad would result from it.

File: api-report/datastore-definitions.api.md [hunk 2]
```diff
@@ -37,6 +37,7 @@ export interface IChannel extends IFluidLoadable {
     // (undocumented)
     readonly owner?: string;
     summarize(fullTree?: boolean, trackState?: boolean): ISummaryTreeWithStats;
+    summarizeAsync?(fullTree?: boolean, trackState?: boolean): Promise<ISummaryTreeWithStats>;
```
Comment (vladsud) [inline api-report/datastore-definitions.api.md:40]: I'm not sure if this is worth attaching in this PR, but we should figure out a way to have consistent naming across all layers. Today, "summarize" means async at IFluidDataStoreChannel layer. We should either have summarize & summarizeAsync, or summarizeSync & summarize (or maybe summarizeAsync & summarizeSync) for sync & async calls respectively.

File: packages/dds/shared-object-base/src/sharedObject.ts [hunk 3]
```diff
@@ -29,13 +29,13 @@ import { ISharedObject, ISharedObjectEvents } from "./types";
...
  */
-export abstract class SharedObject<TEvent extends ISharedObjectEvents = ISharedObjectEvents>
+export abstract class SharedObjectCore<TEvent extends ISharedObjectEvents = ISharedObjectEvents>
     extends EventEmitterWithErrorHandling<TEvent> implements ISharedObject<TEvent> {
     /**
      * @param obj - The thing to check if it is a SharedObject
      * @returns Returns true if the thing is a SharedObject
      */
-    public static is(obj: any): obj is SharedObject {
+    public static is(obj: any): obj is SharedObjectCore {
```
Comment (vladsud) [inline packages/dds/shared-object-base/src/sharedObject.ts:38]: It's not directly related to your changes, but I believe this is wrong - opened https://github.com/microsoft/FluidFramework/issues/8594 to track it

File: api-report/datastore-definitions.api.md (hunk 2 above)
Comment (vladsud) [reply api-report/datastore-definitions.api.md:40]: Actually, looking below, can we go with async summarize() and synchronous getAttachSummary()? It will be consistent with data store naming.

Comment (vladsud) [reply api-report/datastore-definitions.api.md:40]: That said, it maybe harder to do renames (today's summarize() is sync at this layer) due to back-compat requirements. Maybe not - one way or another your PR would need to have back-compat handling.

File: packages/dds/shared-object-base/src/sharedObject.ts [hunk 4]
```diff
@@ -87,15 +82,6 @@ export abstract class SharedObject<TEvent extends ISharedObjectEvents = ISharedO
     }
 
     protected get serializer(): IFluidSerializer {
```
Comment (vladsud) [inline packages/dds/shared-object-base/src/sharedObject.ts:84]: I'm not sure if this method be implemented at this layer at all. I.e. why do we need it & this._serializer?
Maybe we do, but not very clear when it should / can be used.

File: packages/dds/shared-object-base/src/sharedObject.ts [hunk 5]
```diff
@@ -519,3 +447,118 @@ export abstract class SharedObject<TEvent extends ISharedObjectEvents = ISharedO
...
+            0x075 /* "SummarySerializer should be used for serializing data during summary." */);
+        return super.serializer;
+    }
+
+    /**
+     * @param id - The id of the shared object
+     * @param runtime - The IFluidDataStoreRuntime which contains the shared object
+     * @param attributes - Attributes of the shared object
+    */
+    constructor(
```
Comment (vladsud) [inline packages/dds/shared-object-base/src/sharedObject.ts:479]: I believe you can remove constructor and TSC will be happy to find one on SharedObjectCore

File: packages/dds/shared-object-base/src/sharedObject.ts (hunk 1 above)
Comment (vladsud) [reply packages/dds/shared-object-base/src/sharedObject.ts:495]: So I believe the only purpose of this flag is to ensure that nobody uses this.serializer while collecting GC data, as that would miss some info (meaning local serializer would not be used).
I think you are right that this is not interesting for summarize() flow, only for GC collection flow.

And either way, I'd suggest removing it, and instead making this._serializer optional, and simply putting undefined there temporarily while collecting GC (and catching if anyone is touching it).

We can go even further and simply swapping this._serializer with another value while doing GC (and restoring it back one GC is over). I like this option way less as it's way more implicit and will not work for async flows.

Comment (vladsud) [reply packages/dds/shared-object-base/src/sharedObject.ts:495]: RE reentrancy: it mostly does not matter, with one caveat - if this._isSummarizing was a counter, not a boolean. Cause the primary purpose of this flag is not to catch reentrancy (it asserts there is no reentrancy because code dealing with this flag is wrong in presence of reentrancy), the purpose of this flag is to ensure nobody uses this.serializer while GC data is collected, that's it.

File: packages/dds/shared-object-base/src/sharedObject.ts [hunk 6]
```diff
@@ -519,3 +447,118 @@ export abstract class SharedObject<TEvent extends ISharedObjectEvents = ISharedO
...
+        return this.summarize(fullTree, trackState);
+    }
+
+    /**
+     * {@inheritDoc (ISharedObject:interface).getGCData}
+     */
+    public getGCData(fullGC: boolean = false): IGarbageCollectionData {
+        // Set _isSummarizing to true. This flag is used to ensure that we only use SummarySerializer (created in
+        // getGCDataCore) to serialize handles in this object's data.
+        assert(!this._isSummarizing, 0x078 /* "Possible re-entrancy! Summary should not already be in progress." */);
```
Comment (vladsud) [inline packages/dds/shared-object-base/src/sharedObject.ts:523]: I think this code should move into getGCDataCore() because getGCDataCore() is the one who sets up and uses custom serializer

File: packages/runtime/datastore/src/channelContext.ts [hunk 7]
```diff
@@ -83,3 +83,20 @@ export function summarizeChannel(
...
     return summarizeResult;
 }
+
+export async function summarizeChannelAsync(
+    channel: IChannel,
+    fullTree: boolean = false,
+    trackState: boolean = false,
+): Promise<ISummaryTreeWithStats> {
+    let summarizeResult: ISummaryTreeWithStats;
+    if (channel.summarizeAsync !== undefined) {
```
Comment (vladsud) [inline packages/runtime/datastore/src/channelContext.ts:93]: Can you please add comment in this format?
// back-compat: added in version X to support newly added IChannel.summarizeAsync()

Comment (vladsud) [reply packages/runtime/datastore/src/channelContext.ts:93]: If that's the only place where we call into IChannel to summarize, then I think we can do rename here to be consistent with other layers. I.e. we can make summarize() be async and just do
```typesript
summarizeResult = await channel.summarize(fullTree, trackState);
```

If it's old version of code, await of non-promise will return value just fine.

That said, I do not believe we allow mixing versions between data store runtime and DDS, so no back-compat should be necessary (disregard my earlier comments on that front).

The only other place where rename will touch is sync flow

**Review by vladsud — APPROVED** [review]
Looks great! Just minor comments on naming

File: packages/dds/shared-object-base/src/sharedObject.ts (hunk 4 above)
Comment (scarlettjlee) [reply packages/dds/shared-object-base/src/sharedObject.ts:84]: This getter is accessible by derived classes whereas this._serializer is not, in order to prevent its use while garbage collecting which only makes sense in SharedObject. I think the way I've split this functionality across SharedObjectCore and SharedObject makes this more confusing.

File: packages/dds/shared-object-base/src/sharedObject.ts (hunk 6 above)
Comment (scarlettjlee) [reply packages/dds/shared-object-base/src/sharedObject.ts:523]: I agree the serializer functionality should be consolidated in one place. Considering that getGCDataCore is meant to be overridden for complex DDSs, I think it would make more sense to put them both in getGCData and pass the serializer to getGCDataCore. Although it's possible that an implementation of getGCDataCore wouldn't use a serializer at all.

File: packages/test/test-end-to-end-tests/src/test/sharedInterval.spec.ts [hunk 8]
```diff
@@ -776,7 +776,7 @@ describeFullCompat("SharedInterval", (getTestObjectProvider) => {
             assert.equal(
                 mapFrom3.get("nestedKey"), "nestedValue", "Incorrect value in interval collection's shared map");
 
-            const summaryBlob = outerString2.summarize().summary.tree.header as ISummaryBlob;
+            const summaryBlob = (await outerString2.summarize()).summary.tree.header as ISummaryBlob;
```
Comment (scarlettjlee) [inline packages/test/test-end-to-end-tests/src/test/sharedInterval.spec.ts:779]: I'm a bit concerned that the test fails with the following error if it has `outerString2.getAttachSummary` instead. Compat issue?

outerString2.getAttachSummary is not a function
TypeError: outerString2.getAttachSummary is not a function
    at Context.<anonymous> (dist/test/sharedInterval.spec.js:616:46)

File: packages/dds/cell/src/cell.ts [hunk 9]
```diff
@@ -203,30 +199,16 @@ export class SharedCell<T = any> extends SharedObject<ISharedCellEvents<T>>
...
-            ],
-        };
-
-        return tree;
+    protected summarizeCore(serializer: IFluidSerializer, fullTree: boolean): ISummaryTreeWithStats {
+        const content: ICellValue = { value: this.data };
+        const blobContent = serializer.stringify(content, this.handle);
+        const builder = new SummaryTreeBuilder();
+        builder.addBlob(snapshotFileName, blobContent);
+        return builder.getSummaryTree();
```
Comment (DLehenbauer) [inline packages/dds/cell/src/cell.ts:211]: The cleanup to use SummaryTreeBuilder() everywhere is very nice, thank you.

**Review by DLehenbauer — APPROVED** [review]

File: experimental/PropertyDDS/packages/property-dds/src/propertyTree.ts [hunk 10]
```diff
@@ -437,7 +439,7 @@ export class SharedPropertyTree extends SharedObject {
 			snapshot.numChunks = chunks.length;
 		}
 
-		return {
+		return convertToSummaryTreeWithStats({
```
Comment (vladsud) [inline experimental/PropertyDDS/packages/property-dds/src/propertyTree.ts:442]: Not sure if you want to add it to this PR, but glancing through the code, we can remove fullTree argument in convertToSummaryTreeWithStats() / convertToSummaryTree() - ITree here will never have id. Even better would be to probably follow up (with a set of separate issues) to remove all ITree-related code from DDSs and use directly summary interfaces (and kill ITree -> summary conversion functions)

File: experimental/dds/tree/src/generic/GenericSharedTree.ts [hunk 11]
```diff
@@ -303,24 +305,13 @@ export abstract class GenericSharedTree<TChange> extends SharedObject<ISharedTre
...
-				},
-			],
-		};
-
-		return tree;
+	public summarizeCore(serializer: IFluidSerializer, fullTree: boolean): ISummaryTreeWithStats {
+		const blobContent = this.saveSerializedSummary({ serializer });
+		const builder = new SummaryTreeBuilder();
+		builder.addBlob(snapshotFileName, blobContent);
+		return builder.getSummaryTree();
```
Comment (vladsud) [inline experimental/dds/tree/src/generic/GenericSharedTree.ts:314]: Given that same sequence of creation single-blob summary repeats in this PR, maybe create a helper function for that?
I.e., this code to look something like
```
return createSingleBlobSummary(snapshotFileName, this.saveSerializedSummary({ serializer }));
```

File: packages/dds/sequence/src/sequence.ts [hunk 12]
```diff
@@ -452,20 +452,13 @@ export abstract class SharedSegmentSequence<T extends ISegment>
             entries,
         };
 
-        return tree;
```
Comment (vladsud) [inline packages/dds/sequence/src/sequence.ts:455]: Feels a bit inconsistent that in other places with two blobs you converted code to use native summary functionality, while here you left ITree -> Summary conversion

File: packages/dds/shared-object-base/src/sharedObject.ts [hunk 13]
```diff
@@ -519,3 +447,89 @@ export abstract class SharedObject<TEvent extends ISharedObjectEvents = ISharedO
...
+        assert(!this._isGCing, 0x078 /* "Possible re-entrancy! Summary should not already be in progress." */);
+        this._isGCing = true;
+
+        let gcData: IGarbageCollectionData;
+        try {
+            const serializer = new SummarySerializer(
+                this.runtime.channelsRoutingContext,
+                (handle: IFluidHandle) => this.handleDecoded(handle),
+            );
+            gcData = this.getGCDataCore(serializer);
```
Comment (vladsud) [inline packages/dds/shared-object-base/src/sharedObject.ts:503]: I find it a bit weird the usage of serializer arg and return value. It feels like only one of them should be used, i.e.
1. either we pass serializer and expect it to be used, with this function interrogating it after the call to extract GC data.
2. OR we do not pass serializer, and let callers do whatever they want RE how they built up GC info.

Given that all getGCDataCore() implementations seems to do serializer.getSerializedRoutes() dance at the end, I'm thinking we want # 1 with whoever wanting more control using SharedObjectCore class and overwriting getGCData() method instead.

File: packages/test/test-end-to-end-tests/src/test/sharedInterval.spec.ts (hunk 8 above)
Comment (vladsud) [reply packages/test/test-end-to-end-tests/src/test/sharedInterval.spec.ts:779]: This code is wrapped with describeFullCompat() which will try to permutate various versions of container runtime & data stores. You are changing interfaces that are not on compat boundary, but this test code assumes "latest and greatest" by using getAttachSummary() as this API did not exist in previous version of data store (/ dds). But describeFullCompat() will given you a provider that will load old code (old data store code and thus old DDSs) that do not have getAttachSummary().

In other words, there is incompatibility of this test code (assuming latest) with actual data store you are getting.
You would need to either use methods that exists in both (that's what you do with summarize()) or disabled back-compat testing for a while.
You will be able to start leveraging getAttachSummary() in two releases.

File: experimental/PropertyDDS/packages/property-dds/src/propertyTree.ts (hunk 10 above)
Comment (scarlettjlee) [reply experimental/PropertyDDS/packages/property-dds/src/propertyTree.ts:442]: I opened issue #8652

File: packages/dds/sequence/src/sequence.ts (hunk 12 above)
Comment (scarlettjlee) [reply packages/dds/sequence/src/sequence.ts:455]: For anything complicated (like the tree entry above), I used convertToSummaryTreeWithStats rather than converting all the code to create summaries directly. I'll defer that fix-up to #8652 for now.

File: packages/dds/shared-object-base/src/sharedObject.ts (hunk 1 above)
Comment (scarlettjlee) [reply packages/dds/shared-object-base/src/sharedObject.ts]: I tried making serializer optional and putting undefined there while GCing. It doesn't seem any better to me, so I'd rather just keep _isSummarizing/_isGCing.
